tests/Unit/ImporterTest.php

```diff
@@ -3,16 +3,51 @@
 namespace Tests\Unit;
 
 use Tests\TestCase;
+use Illuminate\Http\Request;
+use Illuminate\Http\UploadedFile;
+use Chompy\Http\Controllers\ImportController;
+use Illuminate\Validation\ValidationException;
 
 class ImporterTest extends TestCase
 {
     /**
-     * A basic test example.
+     * Test validation rules for email subscription import.
      *
      * @return void
      */
-    public function testExample()
+    public function testValidationRulesForEmailSubscriptionImport()
     {
-        $this->assertTrue(true);
+        $request = new Request;
+        $file = UploadedFile::fake()->create('importer_test.csv');
+
+        $request->replace([
+            'source-detail' => 'test_opt_in',
+            'topics' => [
+                'community',
+            ],
+        ]);
+
+        $controller = new ImportController();
+
+        $this->expectException(ValidationException::class);
+        $controller->store($request, 'email-subscription');
```

Comment:
This is an interesting way of doing this! What's the advantage to instantiating a controller & request by hand, rather than using Laravel's `$this->post('<url>, $payload)` test method? Is there any risk you could keep some old payload around by mistake by re-using the same `$request` for each step?

Reply:
@DFurnes I decided to do it this way (via the controller) since there is no exposed API endpoint to hit this! This method in the controller is triggered when a user uploads a CSV. 

>  Is there any risk you could keep some old payload around by mistake by re-using the same $request for each step?

Great catch! It looks like all the separate requests weren't actually firing off, just the first assertion in the test. I've separated these into three different tests to make sure all of these scenarios are actually being tested.

Reply:
We should still be able to do this with Laravel's built-in HTTP methods, since there _is_ a HTTP endpoint for this (even though it's using web sessions, rather than an API token)! 

For example, this would accomplish the same thing for this test:

```php
$user = \Chompy\User::forceCreate(['role' => 'admin']);
$response = $this->be($user)->postJson('import/email-subscription', [
    'source-detail' => 'test_opt_in',
    'topics' => [
        'community',
    ],
]);

$response->assertStatus(422);
```

Reply:
That said, definitely not something that needs to happen for merge!

Reply:
ah duh thank you, @DFurnes ! I just updated the tests to have the above since we typically write tests this way in other apps and because I think it is a lot more readable! Ready for another review! (Even though I'm technically not at DS anymore, I can still merge, right? #opensource)!